File: backend-javascript/src/main/resources/org/shedlang/compiler/backends/javascript/modules/Stdlib/Platform/Strings.js
"use strict";
var Options = require("../Options");

function codePointToHexString(char) {
    return char.codePointAt(0).toString(16).toUpperCase();
}

function codePointToInt(char) {
    return BigInt(char.codePointAt(0));
}

function codePointToString(char) {
    return char;
}

function codePointCount(string) {
    let count = 0;
    for (const char of string) {
        count++;
    }
    return BigInt(count);
}

function firstCodePoint(string) {
    if (string.length === 0) {
        return Options.none;
    } else {
        return Options.some(String.fromCodePoint(string.codePointAt(0)))
    }
}

function flatMapCodePoints(func, string) {
    let result = "";
    for (const char of string) {
        result += func(char);
    }
    return result;
}

function foldLeftCodePoints(func, initial, string) {
    let result = initial;
    for (const char of string) {
        result = func(result, char);
    }
    return result;
}

function repeat(string, times) {
    let result = "";
    for (let i = 0n; i < times; i++) {
      result += string;
    }
    return result;
}

function replace(old, replacement, string) {
    return string.split(old).join(replacement);
}

function substring(startIndex, endIndex, string) {
    return string.substring(Number(startIndex), Number(endIndex));
}

exports.codePointToHexString = codePointToHexString;
exports.codePointToInt = codePointToInt;
exports.codePointToString = codePointToString;
exports.codePointCount = codePointCount;
exports.firstCodePoint = firstCodePoint;
exports.flatMapCodePoints = flatMapCodePoints;
exports.foldLeftCodePoints = foldLeftCodePoints;
exports.repeat = repeat;
exports.replace = replace;
exports.substring = substring;

File: backend-python/src/main/resources/org/shedlang/compiler/backends/python/modules/Stdlib/Platform/Strings.py
import io

from .. import Options

def code_point_at(index, string):
    if index < len(string):
        return Options.some(string[index])
    else:
        return Options.none


def code_point_to_hex_string(char):
    return format(ord(char), "X")


def code_point_to_int(char):
    return ord(char)


def code_point_to_string(char):
    return char


def code_point_count(string):
    return len(string)


def flat_map_code_points(func, string):
    result = io.StringIO()
    for char in string:
        result.write(func(char))
    return result.getvalue()


def fold_left_code_points(func, initial, string):
    result = initial
    for char in string:
        result = func(result, char)
    return result


def repeat(string, times):
    return string * times


def replace(old, new, string):
    return string.replace(old, new)


def substring(start_index, end_index, value):
    return value[start_index:end_index]
